lib/line-numbers.js

```diff
@@ -0,0 +1,118 @@
+'use strict';
+
+const callsites = require('callsites');
+const micromatch = require('micromatch');
+const sourceMapSupport = require('source-map-support');
+const flatten = require('lodash/flatten');
+const parseTestSourceInFile = require('./test-ast');
+
+const NUMBER_REGEX = /^-?\d+$/;
+const RANGE_REGEX = /^(?<startGroup>-?\d+)-(?<endGroup>-?\d+)$/;
+const LINE_NUMBERS_REGEX = /^(?:-?\d+(?:--?\d+)?,?)+$/;
+const DELIMITER = ':';
+
+const pluralizeIf = (string, condition) => condition ? `${string}s` : string;
+
+class FormatError extends Error {
+	constructor(offender) {
+		const numberPluralized = pluralizeIf('number', offender && !NUMBER_REGEX.test(offender) && offender.length !== 1);
+		super(`Invalid line ${numberPluralized}: \`${offender}\`. Expected comma-separated list of \`[X|Y-Z]\`.`);
+		this.name = 'FormatError';
+	}
+}
+
+const distinctArray = array => [...new Set(array)];
+const sortNumbersAscending = array => {
+	const sorted = [...array];
+	sorted.sort((a, b) => a - b);
+	return sorted;
+};
+
+const parseNumber = string => parseInt(string, 10);
+const removeAllWhitespace = string => string.replace(/\s/g, '');
+const range = (start, end) => new Array(end - start + 1).fill(start).map((element, index) => element + index);
+
+const parseLineNumbers = suffix => sortNumbersAscending(distinctArray(flatten(suffix.split(',').map(part => {
+	if (NUMBER_REGEX.test(part)) {
+		const number = parseNumber(part);
+		if (number <= 0) {
+			throw new Error(`Invalid line number: \`${part}\`. Line numbers must be positive.`);
+		}
+
+		return number;
+	}
+
+	const match = RANGE_REGEX.exec(part);
+	if (match) {
+		const {startGroup, endGroup} = match.groups;
+		const start = parseNumber(startGroup);
+		const end = parseNumber(endGroup);
+		if (start < 0 || end < 0) {
+			throw new Error(`Invalid line number range: \`${part}\`. Line numbers must be positive.`);
+		}
+
+		if (start > end) {
+			throw new Error(`Invalid line number range: \`${part}\`. \`start\` must be less than \`end\`.`);
+		}
+
+		return range(start, end);
+	}
+
+	throw new FormatError(part);
+}))));
+
+function splitPatternAndLineNumbers(pattern) {
+	const parts = pattern.split(DELIMITER);
+	if (parts.length === 1) {
+		return {pattern, lineNumbers: null};
+	}
+
+	const suffix = removeAllWhitespace(parts.pop());
+	if (!LINE_NUMBERS_REGEX.test(suffix)) {
+		return {pattern, lineNumbers: null};
+	}
+
+	return {pattern: parts.join(DELIMITER), lineNumbers: parseLineNumbers(suffix)};
```

Comment:
On the one hand, this tries to ignore patterns that don't look like they select line numbers. But then on the other hand this throw errors when the numbers or ranges look invalid.

`/^(?:\d+(?:-\d+)?,?)+$/` should match individual numbers and ranges. This just leaves us with a range where the start is after the end. Maybe we should normalize that range?

Anything else would be treated as part of the pattern.

Reply:
Yes, I came to think of this too. Reversed ranges _and_ negative numbers, right?

I don't quite get what you mean by _"normalize that range"_, though? Do you mind clarifying/giving an example?

Also, I don't think `throw new FormatError(...)` can occur/is needed anymore? On the other hand, I don't really know what to do if it were to fall through, so might be better to leave it...

Reply:
The regular expression wouldn't match negative numbers, so it'd be treated as part of the glob pattern instead.

For reverse ranges (start > end) we could normalize the range, so swap start and end.

Now there shouldn't be any errors.

Reply:
> The regular expression wouldn't match negative numbers, so it'd be treated as part of the glob pattern instead.

I beg to differ. Take another look. 

> For reverse ranges (start > end) we could normalize the range, so swap start and end.

Ah, works for me.

Reply:
> > The regular expression wouldn't match negative numbers, so it'd be treated as part of the glob pattern instead.
> 
> I beg to differ. Take another look.

I meant this one: `/^(?:\d+(?:-\d+)?,?)+$/`

`/^(?:-?\d+(?:--?\d+)?,?)+$/` does match negative numbers.

Reply:
Right